pkg/cpio/cpio.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2013-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package cpio
+
+import (
+	"fmt"
+	"io"
+	"log"
+)
+
+var (
+	formatMap map[string]RecordFormat
+	formats   []string
+	Debug     = func(string, ...interface{}) {}
+)
+
+const Trailer = "TRAILER!!!"
+
+var TrailerRecord = Record{nil, Info{Name: Trailer}}
+
+func AddFormat(n string, f RecordFormat) {
+	if _, ok := formatMap[n]; ok {
+		log.Fatalf("cpio: two requests for format %s", n)
+	}
+
+	formatMap[n] = f
+	formats = append(formats, n)
+}
+
+func Format(n string) (RecordFormat, error) {
+	op, ok := formatMap[n]
+	if !ok {
+		return nil, fmt.Errorf("Format %v is not one of %v", n, formats)
+	}
+	return op, nil
+}
+
+func ReadRecord(rr RecordReader) (Record, error) {
+	rec, err := rr.ReadRecord()
+	if err != nil {
+		return Record{}, err
+	}
+	if rec.Name == Trailer {
+		return Record{}, io.EOF
+	}
+	return rec, nil
+}
+
+func ReadRecords(rr RecordReader) ([]Record, error) {
+	var files []Record
+	for {
+		f, err := ReadRecord(rr)
+		if err == io.EOF {
+			return files, nil
+		}
+		if err != nil {
+			return nil, err
+		}
+		files = append(files, f)
+	}
+	return files, nil
+}
+
+func WriteRecords(rw RecordWriter, files []Record) error {
+	for _, f := range files {
+		if err := rw.WriteRecord(f); err != nil {
+			return fmt.Errorf("WriteRecords: writing %q got %v", f.Info.Name, err)
+		}
+	}
+	return WriteFinish(rw)
```

Comment:
let's not put this writefinish here. There are all kinds of reasons not to wire it in: we might want to write a file with no trailer so we can cat them together later as for linux initramfs; we might have different part of a program writing groups of records. . If people want to write a finish record, they can append one to the files list (and we should have a constant TrailerRecord somewhere for that) or they can call WriteFinish explicitly.